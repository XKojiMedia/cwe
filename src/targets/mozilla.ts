import { CweConfig, ManifestIcons, ExtensionInfo, ExtensionCompiler } from '../config';
const webExt = require('web-ext').default;

interface MozillaAddonBrowserAction {
  default_icon?: ManifestIcons;
}
interface MozillaAddonBackground {
  scripts: string[];
}
interface MozillaAddonOptionsUi {
  page: string;
  open_in_tab: boolean;
}

interface MozillaAddon {
  manifest_version: 2;
  name: string;
  short_name: string;
  description: string;
  version: string;
  icons: ManifestIcons;
  browser_action: MozillaAddonBrowserAction;
  permissions: string[];
  content_security_policy: string;
  background: MozillaAddonBackground;
  options_ui?: MozillaAddonOptionsUi;
  offline_enabled: boolean;
}

const buildBrowserAction = (config: CweConfig): MozillaAddonBrowserAction => {
  if (config.manifestOptions?.browserAction) {
    return {
      default_icon: config.manifestOptions.browserAction.defaultIcon,
    };
  }
  return {};
};

const buildOptionsUi = (config: CweConfig): MozillaAddonOptionsUi | undefined => {
  if (config.manifestOptions?.settingsOptions) {
    return {
      page: config.manifestOptions.settingsOptions.page,
      open_in_tab: config.manifestOptions.settingsOptions.openInTab,
    };
  }
  return undefined;
};

const buildExtensionData = (config: CweConfig): MozillaAddon | null => {
  if (!config.manifestOptions) {
    return null;
  }

  return {
    manifest_version: 2,
    name: config.manifestOptions.name,
    short_name: config.manifestOptions.shortName,
    description: config.manifestOptions.description,
    version: config.manifestOptions.version,
    icons: config.manifestOptions.icons,
    browser_action: buildBrowserAction(config),
    permissions: config.manifestOptions.permissions,
    content_security_policy: config.manifestOptions.contentSecurityPolicy,
    background: config.manifestOptions.backgroundOptions,
    options_ui: buildOptionsUi(config),
    offline_enabled: config.manifestOptions.offlineEnabled,
  };
};

export const generateExtensionInfo = async(config: CweConfig): Promise<ExtensionInfo> => {
  return {
    content: JSON.stringify(buildExtensionData(config), null, 2),
    fileName: 'manifest.json',
    fileType: 'json',
  };
};

export default generateExtensionInfo;

export const compileExtension: ExtensionCompiler = async (opts) => {
  return await webExt.cmd.build({
    sourceDir: opts.extensionFilesDir,
    overwriteDest: true,
    artifactsDir: opts.extensionBuildOutputDir,
  });
};
