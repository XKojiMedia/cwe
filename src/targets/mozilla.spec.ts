import { generateExtensionInfo, compileExtension } from './mozilla';
import path from 'path';
import fs from 'fs';
import fsExtra from 'fs-extra';
const webExtModule = require('web-ext');

jest.useFakeTimers();

jest.mock('path');
jest.mock('fs');
jest.mock('fs-extra');

describe('mozilla target', () => {
  describe('generateExtensionInfo', () => {
    it('should generate empty content if no manifest option', async () => {
      const extensionInfo = await generateExtensionInfo({
        exclude: [],
        include: [],
        outDir: '',
        rootDir: '',
        targets: [],
      });
      expect(extensionInfo.content).toBe('');
      expect(extensionInfo.fileName).toBe('manifest.json');
      expect(extensionInfo.fileType).toBe('json');
    });

    it('should generate the manifest content if manifest option provided', async () => {
      const extensionInfo = await generateExtensionInfo({
        exclude: [],
        include: [],
        outDir: '',
        rootDir: '',
        targets: [],
        manifestOptions: {
          name: 'Test extension',
          version: '0.0.1',
          background: {
            scripts: []
          },
          browser_action: {
            default_icon: '',
            default_title: '',
            default_popup: '',
          },
          content_security_policy: '',
          description: 'Test extension description',
          icons: '',
          offline_enabled: false,
          page_action: {
            default_icon: '',
            default_title: '',
            default_popup: '',
          },
          permissions: [],
          options_ui: {
            page: '',
            open_in_tab: false,
          },
          short_name: 'Test'
        }
      });
      expect(extensionInfo.content).toBeTruthy();
      expect(JSON.parse(extensionInfo.content)).toMatchSnapshot();
      expect(extensionInfo.fileName).toBe('manifest.json');
      expect(extensionInfo.fileType).toBe('json');
    });
  });

  describe('compileExtension', () => {
    it('should call web-ext build command with specified parameters', async () => {
      await compileExtension({
        config: {
          exclude: [],
          include: [],
          outDir: '',
          rootDir: '',
          targets: [],
        },
        extensionBuildOutputDir: 'path/to/output',
        extensionFilesDir: 'path/to/files',
      });

      expect(webExtModule.default.cmd.build).toHaveBeenCalledWith({
        sourceDir: 'path/to/files',
        artifactsDir: 'path/to/output',
        overwriteDest: true
      });
    });
  });
});