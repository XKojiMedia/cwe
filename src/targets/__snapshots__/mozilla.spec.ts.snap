// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mozilla target generateExtensionInfo should generate the manifest content if manifest option provided 1`] = `
Object {
  "background": Object {
    "scripts": Array [],
  },
  "browser_action": Object {
    "default_icon": "",
    "default_popup": "",
    "default_title": "",
  },
  "content_security_policy": "",
  "description": "Test extension description",
  "icons": "",
  "manifest_version": 2,
  "name": "Test extension",
  "offline_enabled": false,
  "options_ui": Object {
    "open_in_tab": false,
    "page": "",
  },
  "permissions": Array [],
  "short_name": "Test",
  "version": "0.0.1",
}
`;
